tutorial: java notification app (#352)
localstack/docs: content/en/tutorials/java-notification-app/index.md

---
title: "Building a Java Notification app using AWS Java SDK, Simple Email Service (SES), and CloudFormation"
linkTitle: "Building a Java Notification app using AWS Java SDK, Simple Email Service (SES), and CloudFormation"
weight: 3
description: >
  Build a Java Spring Boot application to configure Simple Email Service (SMS) to send messages using AWS Java SDK. Learn how to configure Simple Queue Service (SQS) & Simple Notification Service (SNS) by using a local CloudFormation infrastructure mocked by LocalStack.
cascade:
  type: docs
---

> AWS Java SDK provides a set of libraries that enables Java developers to work with Amazon Web Services and build applications that work with various AWS services, like Simple Email Service (SES), Simple Queue Service (SQS), Simple Notification Service (SNS), and more. Simple Email Service (SES) is a cloud-based email-sending service that enables developers to integrate email functionality into their applications running on AWS. SES allows developers to work without an on-prem Simple Mail Transfer Protocol (SMTP) system and send bulk emails to many recipients.

Reviewer: ```suggestion
Java is a popular platform for cloud applications that use Amazon Web Services.
With the AWS Java SDK, Java developers can build applications that work with various AWS services, like Simple Email Service (SES), Simple Queue Service (SQS), Simple Notification Service (SNS), and more. Simple Email Service (SES) is a cloud-based email-sending service that enables developers to integrate email functionality into their applications running on AWS. SES allows developers to work without an on-prem Simple Mail Transfer Protocol (SMTP) system and send bulk emails to many recipients.
```

[LocalStack Pro](https://app.localstack.cloud/) supports SES along with a simple user interface to inspect email accounts and sent messages. LocalStack also supports sending SES messages through an actual SMTP email server. We will use Simple Queue Service (SQS) and Simple Notification Service (SNS) to process the emails. We would further employ a CloudFormation stack to configure the infrastructure and configure SNS & SQS subscriptions. AWS Java SDK would be employed to receive these SQS messages and to send these messages through SES further.

> In this tutorial, we will build a Java Spring Boot application to provision a locally emulated infrastructure over LocalStack using CloudFormation, and configure SES, SQS, and SNS to send notifications through Java AWS SDK. We will further use [MailHog](https://github.com/mailhog/MailHog), a local SMTP server, to inspect the emails sent through SES via an intuitive user interface.

Reviewer: ```suggestion
In this tutorial, we will build a Java Spring Boot application that uses locally emulated AWS infrastructure on LocalStack provisioned by CloudFormation, and that uses the Java AWS SDK to send SES, SQS, and SNS messages. We will further use [MailHog](https://github.com/mailhog/MailHog), a local SMTP server, to inspect the emails sent through SES via an intuitive user interface.
```

    "Signature": "EXAMPLEpH+..",
    "Timestamp": "<TIMESTAMP>",
    "SigningCertURL": "https://sns.us-east-1.amazonaws.com/SimpleNotificationService-0000000000000000000000.pem",
    "MessageId": "<MESSAGE-ID>",
  }
]
```

You can now run the `/process` endpoint to send the queued notifications as emails:

{{< command >}}
$ curl -s localhost:8080/process
{{< / command >}}

To check whether the email has been sent, you can query the LocalStack internal SES endpoint using the following command:

{{< command >}}
$ curl -s localhost:4566/_localstack/ses | jq .
{{< / command >}}

You will see an output similar to the following:

```json
{
  "messages": [
    {
      "Id": "<ID>",
      "Timestamp": "<TIMESTAMP>",
      "Region": "us-east-1",
      "Source": "no-reply@localstack.cloud",
      "Destination": {
        "ToAddresses": [
          "alice@example.com"
        ]
      },
      "Subject": "hello",
      "Body": {
        "text_part": "hello world",
        "html_part": null
      }
    }
  ]
}
```

You can also navigate to the MailHog via the user-interface: [`localhost:8025`](http://localhost:8025/) to check out the email.

## Conclusion

> With this simple Spring Boot application, we have demonstrated how you can provision CloudFormation infrastructure for SNS & SQS subscriptionson LocalStack, receive SQS messages, and SES messages with the AWS Java SDK. Using [LocalStack Pro](https://app.localstack.cloud), you can use our Web user interface to view the sent email messages via your LocalStack container. The code for this tutorial can be found in our [LocalStack Pro samples over GitHub](https://github.com/localstack/localstack-pro-samples/tree/master/java-notification-app).

Reviewer: ```suggestion
In this tutorial, we have demonstrated 1) how you can use CloudFormation to provision infrastructure for SNS & SQS subscriptionson on LocalStack, and 2) how to use the AWS Java SDK and Spring Boot to build an application that sends SQS and SES messages. Using [LocalStack Pro](https://app.localstack.cloud), you can use our Web user interface to view the email messages sent by SES. The code for this tutorial can be found in our [LocalStack Pro samples over GitHub](https://github.com/localstack/localstack-pro-samples/tree/master/java-notification-app).
```

```yaml
version: "3.8"

services:
  localstack:
    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
    image: localstack/localstack
    ports:
      - "127.0.0.1:4510-4559:4510-4559"  # external service port range
      - "127.0.0.1:4566:4566"            # LocalStack Edge Proxy
    environment:
      - DEBUG=1
      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR-}
      - HOST_TMP_FOLDER=${TMPDIR:-/tmp/}localstack
      - DOCKER_HOST=unix:///var/run/docker.sock
      - LOCALSTACK_API_KEY=${LOCALSTACK_API_KEY-}
      - SMTP_HOST=smtp:1025
    volumes:
      - "${TMPDIR:-/tmp}/localstack:/tmp/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"

  smtp:
    image: mailhog/mailhog
    ports:
      - "1025"
      - "8025:8025"
```

The above `docker-compose` file will start LocalStack and pull the MailHog image to start the SMTP server (if it doesn't exist yet!) on port `8025`. You can start LocalStack using the following command:

{{< command >}}
$ LOCALSTACK_API_KEY=<your-api-key> docker-compose up -d
{{< / command >}}

> Once LocalStack is started, we can deploy the CloudFormation stack (which might take a few minutes):

Reviewer 2: nit: shouldn't literally take minutes, usually the deployment would be done within a couple of seconds, right?
```suggestion
Once LocalStack is started, we can deploy the CloudFormation stack (which might take a few moments):
```
